_posts/sumtyper.md

```diff
@@ -0,0 +1,406 @@
+---
+layout: post
+title: "Sumtyper"
+---
+
+# Modellering med varianter/sumtyper/sealed classes
+
+Når vi bygger systemer, lager vi modeller av den virkelige verden. Det er mange måter å bygge disse modellene på, og noen av dem introduserer en del problemer vi helst vil være foruten. Her skal vi utforske en teknikk som kalles sumtyper, som kan redusere feilene vi introduserer i modellene våre, og gjøre systemene vi bygger mer robuste.
+
+## Uheldig kompleksitet
+
+Felles for alle systemer vi bygger, er at vi må ha en modell av det vi jobber med i koden. En modell er en forenklet versjon av den virkelige verden. Noen ganger er den veldig forenklet, og fokuserer bare på noen få begreper. Andre ganger har vi behov for å vite mer om alle detaljene, og modellene kan bli ekstremt kompliserte.
+
+En annen form for kompleksitet som dukker opp i modellene våre, er ugyldige tilstander. Dette er et resultat av måten vi har modellert på, og kan føre til at koden vi skriver må ta hensyn til mange tilfeller som aldri kan oppstå i den virkelige verden. Denne typen kompleksitet er uønsket, og fører til at systemene blir vanskeligere og dyrere og vedlikeholde. Det øker også sannsynligheten for feil i koden, og for at vi ender opp i situasjoner som ikke skal være mulig. Denne typen kompleksitet er det Frederick Phillips Brooks Jr. kaller “accidental complexity”: det er kompleksitet som er skapt av oss som bygger systemet, og som ikke finnes i den virkelige verden. I denne artikkelen skal vi kalle det for “uheldig kompleksitet”, og vi skal se på en teknikk som kan begrense det.
+
+## Et eksempel
+
+Tenk det at du skal lage et system for kommunen, som skal levere viktig informasjon til innbyggerne. Kanskje forsøker man først å levere det digitalt, men det er ikke alle som har registrert en digital postkasse, så man må sende brev i stedet. Da må vi vite adressen til mottakeren. Vi lager et første utkast til en modell i Java:
+
+```java
+class Adresse {
+
+  String navn;
+  String veiadresse;
+  String postnummer;
+  String poststed;
+}
+
+```
+
+Denne modellen fungerer for mange av innbyggerne våre, men det viser seg at ikke alle innbyggerne har en vegadresse. Noen har en matrikkeladresse i stedet, som består av et gårdsnummer og et bruksnummer. Man kan ikke sende post til en matrikkeladresse, men vi tenker oss at Posten har en tjeneste som kan oversette en matrikkeladresse til et stedsnavn, så vi ønsker å lagre denne adressen for innbyggerne våre, dersom de har det. Da kan vi utvide modellen til dette:
```

Comment:
Matrikkeladresse er for spesielt. Bytt ut med eksemplene til posten: https://www.posten.no/sende/klargjoring/adressering